src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs

```diff
@@ -94,53 +122,178 @@ internal BlameCollector(IBlameReaderWriter blameReaderWriter, IProcessDumpUtilit
+                if (this.collectProcessDumpCollectionBasedOnTrigger)
+                {
+                    this.ValidateAndAddTriggerBasedProcessDumpParameters(collectDumpNode);
+                }
+
+                var collectHangBasedDumpNode = this.configurationElement[Constants.CollectDumpOnTestSessionHang];
+                this.collectProcessDumpOnTestHostHang = collectHangBasedDumpNode != null;
+                if (this.collectProcessDumpOnTestHostHang)
                 {
-                    this.ValidateAndAddProcessDumpParameters(collectDumpNode);
+                    this.ValidateAndAddHangBasedProcessDumpParameters(collectDumpNode);
                 }
             }
 
             this.attachmentGuid = Guid.NewGuid().ToString().Replace("-", string.Empty);
+
+            if (this.collectProcessDumpOnTestHostHang)
+            {
+                this.inactivityTimer = this.inactivityTimer ?? new InactivityTimer(this.CollectDumpAndAbortTesthost);
+                this.ResetInactivityTimer();
+            }
         }
 
-        private void ValidateAndAddProcessDumpParameters(XmlElement collectDumpNode)
+        /// <summary>
+        /// Disposes of the timer when called to prevent further calls.
+        /// Kills the other instance of procdump if launched for collecting trigger based dumps.
+        /// Starts and waits for a new proc dump process to collect a single dump and then
+        /// kills the testhost process.
+        /// </summary>
+        private void CollectDumpAndAbortTesthost()
         {
-            foreach (XmlAttribute attribute in collectDumpNode.Attributes)
+            EqtTrace.Info(string.Format(CultureInfo.CurrentUICulture, Resources.Resources.InactivityTimeout, (int)this.inactivityTimespan.TotalMinutes));
+            this.inactivityTimerAlreadyFired = true;
+
+            try
+            {
+                EqtTrace.Verbose("Calling dispose on Inactivity timer.");
+                this.inactivityTimer.Dispose();
+            }
+            catch
+            {
+                EqtTrace.Verbose("Inactivity timer is already disposed.");
+            }
+
+            if (this.collectProcessDumpCollectionBasedOnTrigger)
             {
-                if (string.Equals(attribute.Name, Constants.CollectDumpAlwaysKey, StringComparison.OrdinalIgnoreCase))
+                // Detach procdump from the testhost process to prevent testhost process from crashing
+                // if/when we try to kill the existing proc dump process.
+                this.processDumpUtility.DetachFromTargetProcess(this.testHostProcessId);
+            }
+
+            try
+            {
+                this.processDumpUtility.StartHangBasedProcessDump(this.testHostProcessId, this.attachmentGuid, this.GetResultsDirectory(), this.processFullDumpEnabled);
+            }
+            catch (Exception ex)
+            {
+                EqtTrace.Error($"BlameCollector.CollectDumpAndAbortTesthost: Failed with error {ex}");
```

Comment:
Shouldn't we just return here?

Reply:
hmm, fair.

Reply:
actually no, i would still want to go ahead and end the testhost process